plugins/coreboot/fu-coreboot-fmap.c

```diff
@@ -0,0 +1,233 @@
+		return NULL;
+
+	g_debug ("coreboot: FMAP is available");
+
+	// FIXME: Do sanity check
+	return (struct fmap *)tmp;
+}
+
+gchar *
+fu_plugin_coreboot_find_cbfs_active_partition (FuPlugin *plugin, GError **error)
+{
+        gchar *tmp = NULL;
+
+        g_debug ("coreboot: Trying /sys/firmware/cbfs_active_partition");
+
+        if (!g_file_get_contents ("/sys/firmware/cbfs_active_partition", &tmp, NULL, error))
+                return NULL;
+
+        g_debug ("coreboot: cbfs_active_partition is available");
+
+        // FIXME: Do sanity check
+        return tmp;
+}
+
+/* FIXME: The following only works on some x86 platforms that have STRICT_DEV_MEM disabled */
+static GBytes *
+fu_plugin_coreboot_clone_region (FuPlugin *plugin, GError **error, const struct fmap *fmap, const struct fmap_area *fmap_area)
+{
+	gint fd;
+	void *bios_mmio;
+	goffset off;
+	GBytes *tmp;
+	gint size;
+
+	size = (fmap_area->size + getpagesize()) & ~(getpagesize() - 1);
+
+	// on x86 the firmware is memory mapped below 4GiB
+	fd = open ("/dev/mem", O_RDONLY);
+	if (fd < 0) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_FAILED,
+			     "Failed to open /dev/mem");
+		return NULL;
+	}
+
+	off = (1ULL << 32) - fmap->size + fmap_area->offset;
+	g_debug ("Cloning region '%s' %lx @ %x", fmap_area->name, (unsigned long int)off, (unsigned int)size);
+
+	bios_mmio = mmap (NULL, size, PROT_READ, MAP_SHARED, fd, off);
+	if (bios_mmio == MAP_FAILED) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_FAILED,
+			     "Failed to mmap /dev/mem");
+		close (fd);
+		return NULL;
+	}
+	tmp = g_malloc0 (size);
+	if (!tmp) {
+		munmap (bios_mmio, size);
+		close (fd);
+		return NULL;
+	}
+	memcpy (tmp, bios_mmio, size);
+	munmap (bios_mmio, size);
+	close (fd);
+	return tmp;
+}
+
+static FuDevice*
+fu_plugin_coreboot_add_fmap_device (FuPlugin *plugin,
+				   GError **error,
+				   FuDevice *parent,
+				   struct fmap *fmap,
+				   struct fmap_area *fmap_area,
+				   gboolean immutable)
+{
+	g_autofree gchar *sum = NULL;
+	FuDevice *dev;
+	//GBytes *region;
+	g_autofree const gchar *triplet = NULL;
+	g_autofree gchar *inst_id = NULL;
+	g_autofree gchar *name = NULL;
+
+	dev = fu_device_new ();
+	if (!dev)
+		return NULL;
+
+	fu_device_set_vendor (dev, fu_device_get_vendor (parent));
+	fu_device_set_id (dev, fu_device_get_id (parent));
+	name = fu_plugin_coreboot_get_name_for_type (plugin, (const char *)fmap_area->name);
+	if (!name)
+		name = fu_device_get_name (parent);
+	fu_device_set_name (dev, name);
+	fu_device_set_vendor (dev, fu_device_get_vendor (parent));
+
+	sum = g_strdup_printf ("%s, partition '%s'\n", fu_device_get_summary (parent), fmap_area->name);
+	fu_device_set_summary (dev, sum);
+
+	fu_device_add_flag (dev, FWUPD_DEVICE_FLAG_INTERNAL);
+	fu_device_add_icon (dev, "computer");
+	inst_id = g_strdup_printf ("system-fimware-partition-%s\n", fmap_area->name);
+	fu_device_add_instance_id (dev, "system-fimware-partition");
```

Comment:
this particular instance ID I don't think is necessary.  The main reason for adding an instance ID is so that other plugins can reference it for parentage purpose.  In this case you're adding `main-system-firmware` and also have instance IDs already referring to `system-firmware-partition-%s`

Reply:
I need at least on instance id to generate a GUID from, otherwise the device can't be registered, correct?

Reply:
But you already add `inst_id` after this line.  I don't think you need the generic one here specifically.